covid19dp_submission/download_analyses.py

```diff
@@ -127,6 +141,36 @@ def download_files(analyses_array, download_target_dir, processed_analyses_file)
                     os.remove(download_file_path)
 
 
+@retry(exceptions=(UnfinishedBatchError,), logger=logger, tries=4, delay=10, backoff=1.2, jitter=(1, 3))
+def download_files_via_aspera(analyses_array, download_target_dir, processed_analyses_file, ascp, aspera_id_dsa,
+                              batch_size=100):
+    logger.info(f"total number of files to download: {len(analyses_array)}")
+    remaining_analyses_array = []
+    with open(processed_analyses_file, 'a') as open_file:
+        # This copy won't change throughout the iteration
+        for analysis_batch in chunked(copy.copy(analyses_array), batch_size):
+            download_urls = [
+                f"era-fasp@{analysis['submitted_aspera']}" for analysis in analysis_batch
+            ]
+            command = f'{ascp} -i {aspera_id_dsa} -QT -l 300m -P 33001 {" ".join(download_urls)} {download_target_dir}'
+            run_command_with_output(f"Download batch of covid19 data", command)
+
+            for analysis in analysis_batch:
+                expected_output_file = os.path.join(download_target_dir, os.path.basename(analysis['submitted_aspera']))
+                if os.path.exists(expected_output_file):
+                    open_file.write(f"{analysis['analysis_accession']},{analysis['submitted_ftp']}\n")
+                    # WARNING: This will modify the content of the original analysis array allowing the retry to
+                    # only deal with a subset of files to download.
+                    analyses_array.remove(analysis)
+                else:
+                    logger.warn(f"Failed to download {analysis['submitted_aspera']}")
+                    remaining_analyses_array.append(analysis)
```

Comment:
What's the reason for maintaining `remaining_analyses_array` in addition to modifying `analyses_array`?

Reply:
None. I removed `remaining_analyses_array`